src/Stack/Build/Execute.hs

```diff
@@ -593,6 +596,16 @@ singleBuild ac@ActionContext {..} ee@ExecuteEnv {..} task@Task {..} =
             TTLocal lp -> "build" : map T.unpack (Set.toList $ lpComponents lp)
             TTUpstream _ _ -> ["build"]
 
+    when (shouldBuildHaddock eeBuildOpts eeWanted (packageName package) &&
+          packageHasLibrary package &&
```

Comment:
Is this redundant with `packageHasExposedModules`? If not, under what circumstances may it be necessary? If it is redundant, looks like a good place to use `assert`.